### PR TITLE
use safe cast in propagate_constraints

### LINKED ISSUE #11252: Error evaluating clause `where COL_BIGINT < 1e100` (Found by SQLancer-NoREC)
### Describe the bug

When a `BIGINT` typed column is comparing with a large `DOUBLE` literal, it's not evaluated correct if it's in where clause
```
> create table t0(v0 BIGINT);
0 row(s) fetched.
Elapsed 0.003 seconds.

> insert into t0 values (1),(2),(3);
+-------+
| count |
+-------+
| 3     |
+-------+
1 row(s) fetched.
Elapsed 0.003 seconds.

> select * from t0 where v0>1e100;
Arrow error: Cast error: Can't cast value 1.0000000000000002e100 to type Int64
```
But if this comparison clause is in `SELECT` clause, it can be evaluated successfully. Such inconsistency should be a bug.
```
> select v0>1e100 from t0;
+------------------------------------------------------------------------------------------------------------------------+
| t0.v0 > Float64(10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000) |
+------------------------------------------------------------------------------------------------------------------------+
| false                                                                                                                  |
| false                                                                                                                  |
| false                                                                                                                  |
+------------------------------------------------------------------------------------------------------------------------+
3 row(s) fetched.
Elapsed 0.003 seconds.
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

`PostgreSQL` and `DuckDB` can evaluate both queries successfully.

Found by SQLancer https://github.com/apache/datafusion/issues/11030

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #11252

### CODE REVIEW

**Comment by Lordworms** [comment]
seems like a arrow-rs related problem, but we can use safe cast to avoid it temporarily

File: datafusion/sqllogictest/test_files/cast.slt [hunk 1]
```diff
@@ -69,3 +69,16 @@ query ?
...
+statement ok
+create table t0(v0 BIGINT);
+
+statement ok
+insert into t0 values (1),(2),(3);
+
+query I
+select * from t0 where v0>1e100;
+----
+
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/cast.slt:82]: I suggest a test that also shows the filter passing (I wasn't sure if the whole thing was turned into NULL). But I double checked and this query works on this branch:

```suggestion
query I
select * from t0 where v0<1e100;
----
1
2
3

```

**Review by alamb — APPROVED** [review]
Thank you @Lordworms 

I have one small suggestion for additional test coverage but I don't think it is required

File: datafusion/sqllogictest/test_files/cast.slt (hunk 1 above)
Comment (Lordworms) [reply datafusion/sqllogictest/test_files/cast.slt:82]: I'll add that

**Comment by alamb** [comment]
Thanks @Lordworms 🙏

**Pull request merged**